### PR TITLE
Basic functionalities for LSP Server

### PR DESCRIPTION
**Pull request by ankitpriyarup**
This PR provides functionality for: Function Signature, Markdown documentation & Document Link. I will create another PR later on covering rename functionality very soon.

### CODE REVIEW

File: modules/gdscript/language_server/gdscript_extend_parser.cpp [hunk 1]
```diff
@@ -519,6 +519,28 @@ int ExtendGDScriptParser::get_parameter_count(const lsp::Position &p_cursor, lsp
...
+void ExtendGDScriptParser::get_document_link(List<lsp::DocumentLink> *r_links_parameter) {
+
+	int len = lines.size();
+	for (int i = 0; i < len; i++) {
+		int start_pos = lines[i].find("\"");
+		if (start_pos != -1) {
+			int end_pos = lines[i].find("\"", start_pos + 1);
+			String value = lines[i].substr(start_pos + 1, end_pos - start_pos - 1);
+
+			if (value.begins_with("res://") && ResourceLoader::exists(value)) {
```
Comment (Geequlim) [inline modules/gdscript/language_server/gdscript_extend_parser.cpp]: Use FileAccess to check file exists instead

File: modules/gdscript/language_server/gdscript_extend_parser.cpp [hunk 2]
```diff
@@ -519,6 +519,28 @@ int ExtendGDScriptParser::get_parameter_count(const lsp::Position &p_cursor, lsp
...
+				lsp::DocumentLink link;
+				link.range.start.line = i;
+				link.range.start.character = start_pos + 1;
+				link.range.end.line = i;
+				link.range.end.character = end_pos;
+				link.target = value.replace("res://", GDScriptLanguageProtocol::get_singleton()->get_workspace()->root_uri + "/");
```
Comment (Geequlim) [inline modules/gdscript/language_server/gdscript_extend_parser.cpp]: Why not use `GDScriptWorkspace::get_file_uri` here ?

File: modules/gdscript/language_server/gdscript_extend_parser.cpp [hunk 3]
```diff
@@ -519,6 +519,28 @@ int ExtendGDScriptParser::get_parameter_count(const lsp::Position &p_cursor, lsp
 	return count;
 }
 
+void ExtendGDScriptParser::get_document_link(List<lsp::DocumentLink> *r_links_parameter) {
```
Comment (Geequlim) [inline modules/gdscript/language_server/gdscript_extend_parser.cpp]: I'd like to know what the difference between this with `GDScriptWorkspace::get_file_uri` ?

File: modules/gdscript/language_server/gdscript_extend_parser.cpp [hunk 4]
```diff
@@ -488,6 +488,37 @@ String ExtendGDScriptParser::get_text_for_lookup_symbol(const lsp::Position &p_c
...
+			if (closest_bracket_closing_pos > closest_bracket_opening_pos || closest_bracket_opening_pos == -1) {
+				return 0;
+			}
+			for (int j = closest_bracket_opening_pos; j < p_cursor.character; j++) {
+				if (lines[i][j] == ',') {
+					count++;
+				}
+			}
+			begining_position->character = closest_bracket_opening_pos;
+			while (lines[i][begining_position->character] != ' ' && lines[i][begining_position->character] != ',') {
```
[the remark this answers is not in the document]
Comment (ankitpriyarup) [reply modules/gdscript/language_server/gdscript_extend_parser.cpp]: That would be a nice upgrade, tap to cycle through parameters within function. I'm on it

File: modules/gdscript/language_server/gdscript_extend_parser.cpp (hunk 3 above)
Comment (ankitpriyarup) [reply modules/gdscript/language_server/gdscript_extend_parser.cpp]: get_document_link need to find Array of DocumentLink which consist of a range to which the link is applied as well as the uri to the link. I can reuse get_file_uri for finding the uri part within this function

**Comment by ankitpriyarup** [comment]
> For the markdown dump, you'd better move the code out of the `DocData` class to avoid modify the exists engine code.

Should I create another class 'MarkdownDoc' and make it inherit 'DocData' and then create MarkdownDoc object instead of DocData in editor_help.cpp?

Besides that I'll work on tab to cycle through parameters in function signature feature tomorrow. Thanks for the code review.

**Comment by Geequlim** [comment]
> Should I create another class 'MarkdownDoc' and make it inherit 'DocData' and then create MarkdownDoc object instead of DocData in editor_help.cpp?

You can get all the informations from the DocData instance out side. So you can make a tool to read and dump the doc as markdown. You don't need to  inherit 'DocData' as the inheited class won't be instanced by the engine. You can reference the mono module how they generate glue files.

**Comment by Geequlim** [comment]
I'm closing this PR as the base brach is merged into upstream.
You can rebase and submit your code with modified with reviews here to the master branch of upstream.

**Comment by Jason0214** [comment]
For GSoC: this PR has moved to https://github.com/godotengine/godot/pull/31937

**Pull request closed without merging**